rustls/src/conn.rs

```diff
@@ -937,7 +938,7 @@ impl ConnectionCommon {
 
     /// Are we done? i.e., have we processed all received messages,
     /// and received a close_notify to indicate that no new messages
-    /// will arrive?
+    /// will arrive or ran out of bytes on the incoming TLS stream?
```

Comment:
the comments https://github.com/ctz/rustls/issues/563#issuecomment-799827636, https://github.com/ctz/rustls/issues/563#issuecomment-811446138, https://github.com/ctz/rustls/issues/563#issuecomment-811562250 suggest that a TLS connection closing without a `CloseNotify` ought to be considered an error (although the user should be able to ignore that error if they choose to). it *looks* like this code will now always return a clean error (`Ok(0)`) regardless of whether a `CloseNotify` was received or the TCP connection just closed without one?

I think we should probably be distinguishing between `CloseNotify` and closing a connection without a `CloseNotify`, and return an `io::Error` in that case. 

IMO, [`ErrorKind::UnexpectedEof`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.UnexpectedEof) is probably the most correct error kind to use to indicate that a connection was closed without a `CloseNotify`, but I'm open to being convinced otherwise...i do think it's important to use an `ErrorKind` that rustls won't return in other cases, so user code can programmatically distinguish between the underlying connection EOF and other errors.

I wonder if it's also worth adding an option to consider TCP connection EOFs without `CloseNotify` a clean EOF, since some use cases may require talking to peers that don't support `CloseNotify`? Or is it sufficient to just allow the user to ignore that error kind themself?

Reply:
I agree that this should be reported as an error rather than a 'clean' EOF. UnexpectedEof sounds like an excellent option, as does ConnectionAborted.

Reply:
IMO, I have a slight preference for `UnexpectedEof` over `ConnectionAborted`, as I think it's possible that the I/O reader passed to `read_tls` may return `ConnectionAborted`, but it probably won't return `UnexpectedEof`, so it's clearer where the error came from. But, either is probably fine here.